pandas/core/internals/concat.py

```diff
@@ -283,8 +291,16 @@ def _concat_managers_axis0(
             nb._mgr_locs = nb._mgr_locs.add(offset)
             blocks.append(nb)
 
+        if not made_copy and not copy and using_copy_on_write():
+            refs.extend([weakref.ref(blk) for blk in mgr.blocks])
+            parent = mgr
+
         offset += len(mgr.items)
-    return BlockManager(tuple(blocks), axes)
+
+    result = BlockManager(tuple(blocks), axes)
+    result.parent = parent
+    result.refs = cast(Optional[List[Optional[weakref.ref]]], refs) if refs else None
```

Comment:
Can you also pass this to the BlockManager(..) constructor (that will ensure it is verified, although that should in theory not be needed here)

Reply:
and for the the type annotation, in managers.py we seem to generally use `list[weakref.ref | None] | None`

Reply:
The cast annotation should be equivalent to what you wrote

Reply:
Yep, I know it's equivalent, but I think we should prefer consistent syntax for the same thing

Reply:
Ah got you, but as far as I know: you can’t use list[x, None] in cast, you have to use optional